File: app/server/backend/routers/checklist.py
"""孕程记 - 清单 API。"""

import json
import uuid
from pathlib import Path
from typing import List

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.models.checklist import Checklist, ChecklistItem
from backend.schemas.common import ApiResponse
from backend.schemas.checklist import (
    ChecklistItemCreate, ChecklistItemUpdate, ChecklistItemResponse, ChecklistResponse,
)
from backend.config import settings

router = APIRouter()

# 默认清单配置：四种类型及其对应的数据文件和显示名称
_DEFAULT_CHECKLISTS = [
    {"type": "delivery_bag", "name": "待产包清单", "file": "default_checklist_delivery.json"},
    {"type": "newborn_prep", "name": "新生儿准备清单", "file": "default_checklist_newborn.json"},
    {"type": "delivery_check", "name": "产房待检清单", "file": "default_checklist_delivery_check.json"},
    {"type": "confinement", "name": "月子物品清单", "file": "default_checklist_confinement.json"},
]


async def _ensure_default_checklists(pregnancy_id: str, db: AsyncSession) -> List[Checklist]:
    """确保存在默认清单，如果不存在则自动创建。"""
    result = await db.execute(
        select(Checklist).where(Checklist.pregnancy_id == pregnancy_id)
    )
    existing = result.scalars().all()

    if len(existing) > 0:
        return existing

    created_lists = []
    for cfg in _DEFAULT_CHECKLISTS:
        checklist = Checklist(
            id=str(uuid.uuid4()),
            pregnancy_id=pregnancy_id,
            type=cfg["type"],
            name=cfg["name"],
        )
        db.add(checklist)
        await db.flush()

        filepath = Path(settings.DATA_DIR) / cfg["file"]
        if filepath.exists():
            with open(filepath, "r", encoding="utf-8") as f:
                default_data = json.load(f)

            sort_order = 0
            for cat in default_data.get("categories", []):
                category_name = cat.get("category", "")
                for item_data in cat.get("items", []):
                    if isinstance(item_data, str):
                        item_name = item_data
                        item_desc = None
                        item_mandatory = 0
                    else:
                        item_name = item_data.get("name", "")
                        item_desc = item_data.get("description")
                        item_mandatory = 1 if item_data.get("mandatory", False) else 0
                    item = ChecklistItem(
                        id=str(uuid.uuid4()),
                        checklist_id=checklist.id,
                        name=item_name,
                        description=item_desc,
                        category=category_name,
                        is_checked=0,
                        is_custom=0,
                        is_mandatory=item_mandatory,
                        sort_order=sort_order,
                    )
                    db.add(item)
                    sort_order += 1

        created_lists.append(checklist)

    await db.flush()
    return created_lists


@router.get("/checklists/progress", response_model=ApiResponse)
async def get_checklist_progress(pregnancy_id: str, db: AsyncSession = Depends(get_db)):
    """获取清单完成进度汇总。"""
    result = await db.execute(
        select(Checklist).where(Checklist.pregnancy_id == pregnancy_id)
    )
    checklists = result.scalars().all()

    total_items = 0
    checked_items = 0
    checklist_details = []

    for cl in checklists:
        result = await db.execute(
            select(ChecklistItem).where(ChecklistItem.checklist_id == cl.id)
        )
        items = result.scalars().all()
        cl_total = len(items)
        cl_checked = sum(1 for i in items if i.is_checked == 1)
        total_items += cl_total
        checked_items += cl_checked
        checklist_details.append({
            "id": cl.id,
            "name": cl.name,
            "type": cl.type,
            "total": cl_total,
            "checked": cl_checked,
            "percentage": round(cl_checked / cl_total * 100) if cl_total > 0 else 0,
        })

    return ApiResponse(data={
        "total": total_items,
        "checked": checked_items,
        "percentage": round(checked_items / total_items * 100) if total_items > 0 else 0,
        "checklists": checklist_details,
    })


@router.get("/checklists", response_model=ApiResponse[List[ChecklistResponse]])
async def list_checklists(pregnancy_id: str, db: AsyncSession = Depends(get_db)):
    """清单列表（自动初始化默认清单）。"""
    checklists = await _ensure_default_checklists(pregnancy_id, db)
    return ApiResponse(data=[ChecklistResponse.model_validate(c) for c in checklists])


@router.get("/checklists/{checklist_id}", response_model=ApiResponse[ChecklistResponse])
async def get_checklist(checklist_id: str, db: AsyncSession = Depends(get_db)):
    """清单详情（含条目）。"""
    checklist = await db.get(Checklist, checklist_id)
    if not checklist:
        return ApiResponse(code=1001, data=None, message="清单不存在")
    return ApiResponse(data=ChecklistResponse.model_validate(checklist))


@router.put("/checklists/{checklist_id}", response_model=ApiResponse[ChecklistResponse])
async def update_checklist(checklist_id: str, name: str = None, db: AsyncSession = Depends(get_db)):
    """更新清单。"""
    checklist = await db.get(Checklist, checklist_id)
    if not checklist:
        return ApiResponse(code=1001, data=None, message="清单不存在")
    if name:
        checklist.name = name
    await db.flush()
    await db.refresh(checklist)
    return ApiResponse(data=ChecklistResponse.model_validate(checklist))


@router.post("/checklists/{checklist_id}/items", response_model=ApiResponse[ChecklistItemResponse])
async def add_checklist_item(checklist_id: str, data: ChecklistItemCreate, db: AsyncSession = Depends(get_db)):
    """添加清单条目。"""
    # 获取当前最大排序号
    result = await db.execute(
        select(ChecklistItem).where(ChecklistItem.checklist_id == checklist_id).order_by(ChecklistItem.sort_order.desc())
    )
    last_item = result.scalars().first()
    sort_order = (last_item.sort_order + 1) if last_item else 0

    item = ChecklistItem(
        checklist_id=checklist_id,
        name=data.name,
        description=data.description,
        category=data.category,
        is_custom=data.is_custom,
        is_mandatory=data.is_mandatory,
        sort_order=sort_order,
    )
    db.add(item)
    await db.flush()
    await db.refresh(item)
    return ApiResponse(data=ChecklistItemResponse.model_validate(item))


@router.put("/checklists/items/{item_id}", response_model=ApiResponse[ChecklistItemResponse])
async def update_checklist_item(item_id: str, data: ChecklistItemUpdate, db: AsyncSession = Depends(get_db)):
    """更新条目（勾选/改名）。"""
    item = await db.get(ChecklistItem, item_id)
    if not item:
        return ApiResponse(code=1001, data=None, message="条目不存在")
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(item, key, value)
    await db.flush()
    await db.refresh(item)
    return ApiResponse(data=ChecklistItemResponse.model_validate(item))


@router.delete("/checklists/items/{item_id}", response_model=ApiResponse)
async def delete_checklist_item(item_id: str, db: AsyncSession = Depends(get_db)):
    """删除条目。"""
    item = await db.get(ChecklistItem, item_id)
    if not item:
        return ApiResponse(code=1001, data=None, message="条目不存在")
    await db.delete(item)
    return ApiResponse(message="删除成功")


@router.post("/checklists/{checklist_id}/init-default", response_model=ApiResponse[ChecklistResponse])
async def init_default_checklist(checklist_id: str, db: AsyncSession = Depends(get_db)):
    """初始化默认清单。"""
    checklist = await db.get(Checklist, checklist_id)
    if not checklist:
        return ApiResponse(code=1001, data=None, message="清单不存在")

    # 根据类型加载默认数据
    type_to_file = {
        "delivery_bag": "default_checklist_delivery.json",
        "newborn_prep": "default_checklist_newborn.json",
        "delivery_check": "default_checklist_delivery_check.json",
        "confinement": "default_checklist_confinement.json",
    }

    filename = type_to_file.get(checklist.type)
    if not filename:
        return ApiResponse(code=1002, data=None, message="不支持的清单类型")

    filepath = Path(settings.DATA_DIR) / filename
    if not filepath.exists():
        return ApiResponse(code=2002, data=None, message="默认数据文件不存在")

    with open(filepath, "r", encoding="utf-8") as f:
        default_data = json.load(f)

    # 清空现有条目
    result = await db.execute(select(ChecklistItem).where(ChecklistItem.checklist_id == checklist_id))
    for item in result.scalars().all():
        await db.delete(item)

    # 插入默认条目
    sort_order = 0
    categories = default_data.get("categories", [])
    for cat in categories:
        category_name = cat.get("category", "")
        for item_data in cat.get("items", []):
            if isinstance(item_data, str):
                item_name = item_data
                item_desc = None
                item_mandatory = 0
            else:
                item_name = item_data.get("name", "")
                item_desc = item_data.get("description")
                item_mandatory = 1 if item_data.get("mandatory", False) else 0
            item = ChecklistItem(
                id=str(uuid.uuid4()),
                checklist_id=checklist_id,
                name=item_name,
                description=item_desc,
                category=category_name,
                is_checked=0,
                is_custom=0,
                is_mandatory=item_mandatory,
                sort_order=sort_order,
            )
            db.add(item)
            sort_order += 1

    await db.flush()
    await db.refresh(checklist)
    return ApiResponse(data=ChecklistResponse.model_validate(checklist))

File: app/server/node/routes/checklist.js
const express = require('express');
const router = express.Router();
const path = require('path');
const fs = require('fs');
const db = require('../db');
const config = require('../config');
const logger = require('../logger');

const DEFAULT_CHECKLISTS = [
  { type: 'delivery_bag', name: '待产包清单', file: 'default_checklist_delivery.json' },
  { type: 'newborn_prep', name: '新生儿准备清单', file: 'default_checklist_newborn.json' },
  { type: 'delivery_check', name: '产房待检清单', file: 'default_checklist_delivery_check.json' },
  { type: 'confinement', name: '月子物品清单', file: 'default_checklist_confinement.json' }
];

function _loadDefaultItems(filePath) {
  const raw = fs.readFileSync(filePath, 'utf-8');
  const data = JSON.parse(raw);
  const items = [];
  let sortOrder = 0;
  const categories = data.categories || [];
  for (const cat of categories) {
    const categoryName = cat.category || '';
    for (const itemData of (cat.items || [])) {
      items.push({
        name: itemData.name || '',
        description: itemData.description || null,
        category: categoryName,
        is_mandatory: itemData.mandatory ? 1 : 0,
        sort_order: sortOrder++
      });
    }
  }
  return items;
}

function _ensureDefaultChecklists(pregnancyId) {
  const existing = db.queryAll(
    'SELECT id, type FROM checklist WHERE pregnancy_id = ?',
    [pregnancyId]
  );

  const existingTypes = new Set(existing.map(cl => cl.type));
  let createdCount = 0;

  for (const cl of DEFAULT_CHECKLISTS) {
    if (existingTypes.has(cl.type)) continue;

    const filePath = path.join(config.DATA_DIR, cl.file);
    if (!fs.existsSync(filePath)) {
      logger.warn('checklist', `_ensureDefaultChecklists - file not found: ${cl.file}`);
      continue;
    }

    let items;
    try {
      items = _loadDefaultItems(filePath);
    } catch (e) {
      logger.warn('checklist', `Failed to load ${cl.file}: ${e.message}`);
      continue;
    }

    const checklistId = db.generateId();
    db.run(
      'INSERT INTO checklist (id, pregnancy_id, type, name) VALUES (?, ?, ?, ?)',
      [checklistId, pregnancyId, cl.type, cl.name]
    );

    items.forEach((item) => {
      const itemId = db.generateId();
      db.run(
        `INSERT INTO checklist_item (id, checklist_id, name, description, category, is_checked, is_custom, is_mandatory, sort_order)
         VALUES (?, ?, ?, ?, ?, 0, 0, ?, ?)`,
        [itemId, checklistId, item.name, item.description, item.category, item.is_mandatory, item.sort_order]
      );
    });

    logger.info('checklist', `_ensureDefaultChecklists - created "${cl.name}" with ${items.length} items for pregnancy_id=${pregnancyId}`);
    createdCount++;
  }

  if (createdCount > 0) {
    logger.info('checklist', `_ensureDefaultChecklists - created ${createdCount} default checklists for pregnancy_id=${pregnancyId}`);
  }
}

router.get('/checklists/progress', (req, res) => {
  try {
    const { pregnancy_id } = req.query;
    if (!pregnancy_id) {
      logger.warn('checklist', 'GET /checklists/progress - missing pregnancy_id');
      return res.json({ code: 1001, data: null, message: 'pregnancy_id 为必填项' });
    }
    logger.info('checklist', `GET /checklists/progress - pregnancy_id=${pregnancy_id}`);

    _ensureDefaultChecklists(pregnancy_id);

    const checklists = db.queryAll(
      `SELECT c.id, c.name, c.type,
              (SELECT COUNT(*) FROM checklist_item WHERE checklist_id = c.id) as total,
              (SELECT COUNT(*) FROM checklist_item WHERE checklist_id = c.id AND is_checked = 1) as checked
       FROM checklist c
       WHERE c.pregnancy_id = ?
       ORDER BY c.created_at`,
      [pregnancy_id]
    );

    let totalItems = 0;
    let checkedItems = 0;
    const list = [];

    for (const cl of checklists) {
      const t = cl.total || 0;
      const ck = cl.checked || 0;
      totalItems += t;
      checkedItems += ck;
      list.push({
        id: cl.id,
        name: cl.name,
        type: cl.type,
        total: t,
        checked: ck,
        percentage: t > 0 ? Math.round((ck / t) * 100) : 0
      });
    }

    logger.info('checklist', `GET /checklists/progress - ${list.length} checklists, total=${totalItems}, checked=${checkedItems}`);
    res.json({
      code: 0,
      data: {
        total: totalItems,
        checked: checkedItems,
        percentage: totalItems > 0 ? Math.round((checkedItems / totalItems) * 100) : 0,
        checklists: list
      },
      message: 'success'
    });
  } catch (e) {
    logger.error('checklist', `GET /checklists/progress error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/checklists', (req, res) => {
  try {
    const { pregnancy_id } = req.query;
    if (!pregnancy_id) {
      logger.warn('checklist', 'GET /checklists - missing pregnancy_id');
      return res.json({ code: 1001, data: null, message: 'pregnancy_id 为必填项' });
    }
    logger.info('checklist', `GET /checklists - pregnancy_id=${pregnancy_id}`);

    _ensureDefaultChecklists(pregnancy_id);

    const checklists = db.queryAll(
      'SELECT * FROM checklist WHERE pregnancy_id = ? ORDER BY created_at',
      [pregnancy_id]
    );

    for (const cl of checklists) {
      cl.items = db.queryAll(
        'SELECT * FROM checklist_item WHERE checklist_id = ? ORDER BY sort_order',
        [cl.id]
      );
    }

    logger.info('checklist', `GET /checklists - ${checklists.length} checklists with items`);
    res.json({ code: 0, data: checklists, message: 'success' });
  } catch (e) {
    logger.error('checklist', `GET /checklists error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/checklists/:id', (req, res) => {
  try {
    logger.info('checklist', `GET /checklists/${req.params.id}`);
    const checklist = db.queryOne('SELECT * FROM checklist WHERE id = ?', [req.params.id]);
    if (!checklist) {
      logger.warn('checklist', `GET /checklists/${req.params.id} - not found`);
      return res.json({ code: 1001, data: null, message: '清单不存在' });
    }

    const items = db.queryAll(
      'SELECT * FROM checklist_item WHERE checklist_id = ? ORDER BY sort_order, created_at',
      [req.params.id]
    );

    logger.info('checklist', `GET /checklists/${req.params.id} - found, ${items.length} items`);
    res.json({ code: 0, data: { ...checklist, items }, message: 'success' });
  } catch (e) {
    logger.error('checklist', `GET /checklists/:id error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.put('/checklists/:id', (req, res) => {
  try {
    logger.info('checklist', `PUT /checklists/${req.params.id} - name=${req.body.name}`);
    const existing = db.queryOne('SELECT * FROM checklist WHERE id = ?', [req.params.id]);
    if (!existing) {
      logger.warn('checklist', `PUT /checklists/${req.params.id} - not found`);
      return res.json({ code: 1001, data: null, message: '清单不存在' });
    }

    const { name } = req.body;
    if (!name) {
      return res.json({ code: 1001, data: null, message: 'name 为必填项' });
    }

    db.run("UPDATE checklist SET name = ? WHERE id = ?", [name, req.params.id]);

    const updated = db.queryOne('SELECT * FROM checklist WHERE id = ?', [req.params.id]);
    logger.info('checklist', `PUT /checklists/${req.params.id} - updated`);
    res.json({ code: 0, data: updated, message: 'success' });
  } catch (e) {
    logger.error('checklist', `PUT /checklists/:id error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.post('/checklists/:id/items', (req, res) => {
  try {
    logger.info('checklist', `POST /checklists/${req.params.id}/items - name=${req.body.name}`);
    const existing = db.queryOne('SELECT * FROM checklist WHERE id = ?', [req.params.id]);
    if (!existing) {
      logger.warn('checklist', `POST /checklists/${req.params.id}/items - checklist not found`);
      return res.json({ code: 1001, data: null, message: '清单不存在' });
    }

    const { name, description, category, is_custom, is_mandatory } = req.body;
    if (!name) {
      return res.json({ code: 1001, data: null, message: 'name 为必填项' });
    }

    const maxOrder = db.queryOne(
      'SELECT COALESCE(MAX(sort_order), -1) as max_order FROM checklist_item WHERE checklist_id = ?',
      [req.params.id]
    );

    const itemId = db.generateId();
    db.run(
      `INSERT INTO checklist_item (id, checklist_id, name, description, category, is_checked, is_custom, is_mandatory, sort_order)
       VALUES (?, ?, ?, ?, ?, 0, ?, ?, ?)`,
      [itemId, req.params.id, name, description || null, category || '', is_custom ? 1 : 0, is_mandatory ? 1 : 0, maxOrder.max_order + 1]
    );

    const item = db.queryOne('SELECT * FROM checklist_item WHERE id = ?', [itemId]);
    logger.info('checklist', `POST /checklists/${req.params.id}/items - item created (id=${itemId})`);
    res.json({ code: 0, data: item, message: 'success' });
  } catch (e) {
    logger.error('checklist', `POST /checklists/:id/items error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.put('/checklists/items/:item_id', (req, res) => {
  try {
    logger.info('checklist', `PUT /checklists/items/${req.params.item_id} - fields=${Object.keys(req.body).join(',')}`);
    const existing = db.queryOne('SELECT * FROM checklist_item WHERE id = ?', [req.params.item_id]);
    if (!existing) {
      logger.warn('checklist', `PUT /checklists/items/${req.params.item_id} - not found`);
      return res.json({ code: 1001, data: null, message: '条目不存在' });
    }

    const updates = [];
    const params = [];

    if (req.body.name !== undefined) {
      updates.push('name = ?');
      params.push(req.body.name);
    }
    if (req.body.description !== undefined) {
      updates.push('description = ?');
      params.push(req.body.description);
    }
    if (req.body.is_checked !== undefined) {
      updates.push('is_checked = ?');
      params.push(req.body.is_checked ? 1 : 0);
    }
    if (req.body.is_mandatory !== undefined) {
      updates.push('is_mandatory = ?');
      params.push(req.body.is_mandatory ? 1 : 0);
    }
    if (req.body.category !== undefined) {
      updates.push('category = ?');
      params.push(req.body.category);
    }
    if (req.body.sort_order !== undefined) {
      updates.push('sort_order = ?');
      params.push(req.body.sort_order);
    }

    if (updates.length === 0) {
      return res.json({ code: 1001, data: null, message: '没有需要更新的字段' });
    }

    params.push(req.params.item_id);
    db.run(`UPDATE checklist_item SET ${updates.join(', ')} WHERE id = ?`, params);

    const updated = db.queryOne('SELECT * FROM checklist_item WHERE id = ?', [req.params.item_id]);
    logger.info('checklist', `PUT /checklists/items/${req.params.item_id} - updated`);
    res.json({ code: 0, data: updated, message: 'success' });
  } catch (e) {
    logger.error('checklist', `PUT /checklists/items/:id error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.delete('/checklists/items/:item_id', (req, res) => {
  try {
    logger.info('checklist', `DELETE /checklists/items/${req.params.item_id}`);
    const existing = db.queryOne('SELECT * FROM checklist_item WHERE id = ?', [req.params.item_id]);
    if (!existing) {
      logger.warn('checklist', `DELETE /checklists/items/${req.params.item_id} - not found`);
      return res.json({ code: 1001, data: null, message: '条目不存在' });
    }

    db.run('DELETE FROM checklist_item WHERE id = ?', [req.params.item_id]);
    logger.info('checklist', `DELETE /checklists/items/${req.params.item_id} - deleted`);
    res.json({ code: 0, data: null, message: '删除成功' });
  } catch (e) {
    logger.error('checklist', `DELETE /checklists/items/:id error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.post('/checklists/:id/init-default', (req, res) => {
  try {
    logger.info('checklist', `POST /checklists/${req.params.id}/init-default`);
    const checklist = db.queryOne(
      'SELECT * FROM checklist WHERE id = ?',
      [req.params.id]
    );
    if (!checklist) {
      logger.warn('checklist', `POST /checklists/${req.params.id}/init-default - not found`);
      return res.json({ code: 1001, data: null, message: '清单不存在' });
    }

    db.run('DELETE FROM checklist_item WHERE checklist_id = ?', [req.params.id]);

    const defaultCl = DEFAULT_CHECKLISTS.find(c => c.type === checklist.type);
    if (!defaultCl) {
      logger.warn('checklist', `POST /checklists/${req.params.id}/init-default - no default config for type=${checklist.type}`);
      return res.json({ code: 1001, data: null, message: '未找到对应的默认配置' });
    }

    const filePath = path.join(config.DATA_DIR, defaultCl.file);
    if (!fs.existsSync(filePath)) {
      logger.warn('checklist', `POST /checklists/${req.params.id}/init-default - file not found: ${defaultCl.file}`);
      return res.json({ code: 1001, data: null, message: '默认配置文件不存在' });
    }

    let items;
    try {
      items = _loadDefaultItems(filePath);
    } catch (e) {
      logger.error('checklist', `POST /checklists/${req.params.id}/init-default - failed to load config: ${e.message}`);
      return res.json({ code: 1001, data: null, message: '读取默认配置失败: ' + e.message });
    }

    items.forEach((item) => {
      const itemId = db.generateId();
      db.run(
        `INSERT INTO checklist_item (id, checklist_id, name, description, category, is_checked, is_custom, is_mandatory, sort_order)
         VALUES (?, ?, ?, ?, ?, 0, 0, ?, ?)`,
        [itemId, req.params.id, item.name, item.description, item.category, item.is_mandatory, item.sort_order]
      );
    });

    const refreshed = db.queryAll(
      'SELECT * FROM checklist_item WHERE checklist_id = ? ORDER BY sort_order',
      [req.params.id]
    );

    const updatedChecklist = db.queryOne('SELECT * FROM checklist WHERE id = ?', [req.params.id]);
    logger.info('checklist', `POST /checklists/${req.params.id}/init-default - restored ${refreshed.length} items`);
    res.json({ code: 0, data: { ...updatedChecklist, items: refreshed }, message: 'success' });
  } catch (e) {
    logger.error('checklist', `POST /checklists/:id/init-default error: ${e.message}`);
    res.json({ code: 1001, data: null, message: e.message });
  }
});

module.exports = router;
